tests/test_extension_robotstxtparser.py

```diff
@@ -0,0 +1,143 @@
+from twisted.trial import unittest
+from scrapy.utils.python import to_native_str
+from scrapy.extensions.robotstxtparser import PythonRobotParser, ReppyRobotParser, RerpRobotParser
+
+class PythonRobotParserTest(unittest.TestCase):
+    def test_allowed(self):
+        robotstxt_content = ("User-agent: * \n"
```

Comment:
Code for most of the tests methods copied across test classes:
- `test_allowed` is a full copy
- `test_preferred_host` is a full copy
- `test_crawl_delay` is almost a full copy
Can you use some mechanism to write common code only once?

Reply:
I have moved all the test cases to a base class. Then I have used a separate class for each parser, to call relevant test cases. Would this help?